### DISCUSSION TITLE
Runtime folder

### DISCUSSION THREAD

**Discussion by qwertydelle** [pintariching/rustle#66] [Q&A]
What is the point of the runtime folder?

**Comment by pintariching** [comment]
I was playing around and looking where to put the runtime files. The original compiler has them all in a seperate [runtime](https://github.com/sveltejs/svelte/tree/master/src/runtime) folder and I guess the required functions then get copied over to the client.

I was thinking of having a hashmap or something of all the various functions and then you insert whichever functions you need into a `utils.js` and that gets sent along the other files.

**Reply by qwertydelle** [reply]
I don't think the runtime folder will be needed.  Since svelte already provides a runtime. Also from what I have seen while reading svelte's source code the runtime library should be provided as a Path to the compiler instead of the compiler having to inject a runtime.

**Reply by pintariching** [reply]
Doesn't that make us dependent on the Svelte compiler?

I think we need to copy the entire runtime folder and reference it in the compiled output imports?

**Reply by qwertydelle** [reply]
That will make us dependent on the runtime. I think copying the runtime folder is not necessary since it is provided in node_modules whenever working with svelte.

**Reply by pintariching** [reply]
That will also make us dependent on Svelte will it not? And what about if you want to work with Svelte outside of NPM?

**Reply by qwertydelle** [reply]
The user can still provide the runtime at least that's how svelte does it. I found this in the svelte docs [Svelte Compiler function](https://github.com/sveltejs/svelte/edit/master/site/content/docs/05-compile-time.md). One of the arguments for the compiler is sveltePath which I am guessing is where the runtime is located. But also the svelte compiler is built to be integrated  with bundler as opposed to being standalone.

**Reply by pintariching** [reply]
I'm still unsure. The compiler has to have a runtime available so the bundler can do it's job. 

Yes we could reference the original svelte runtime and that would require dependencies to it. If we want to make rustle a substitute for svelte, I don't think we can do that.

**Reply by qwertydelle** [reply]
I thought rustle was suppose to replace the compiler part of svelte not the runtime and the compiler. At that point won't it be more of its own thing rather than a svelte compiler?

**Reply by pintariching** [reply]
Without copying over the runtime from svelte, we cannot have our compiled code working. I first thought to assemble the runtime in the compile step but I think letting the bundler do that might be a better idea.

They call it "runtime" but it's basically just convenience functions to make the compiled files more readable.

**Reply by qwertydelle** [reply]
If we really wanted rustle to be stand-alone we could just ship it with a bundled version of the runtime.

**Reply by pintariching** [reply]
Yeah I think we should do that. Just so that the compiled files can be run in a browser without needing to include NPM. One of the main goals of Rustle is to be able to run it without NPM.

**Reply by qwertydelle** [reply]
Sounds good.
